### PR TITLE
Fix multiple bugs

### PR DESCRIPTION
**Pull request by pulsejet**
1. Older articles showed up on refresh, since the existing buffer was
not cleared
2. Same articles would show up sometimes while scrolling since order of
loading was not guaranteed.
3. Repetitive calls were being made when the user reached bottom, making it
unreadable due to too many re-renders (also related to 2)
4. Upgrade some calls for Pie
5. Refactor itemclicklisteners into adapters
6.  Use memory caching for feed and explore
7. Load low resolution placeholders while big images load

### CODE REVIEW

File: app/src/main/java/app/insti/fragment/FeedFragment.java [hunk 1]
```diff
@@ -92,23 +99,20 @@ public void onPause()
...
+        if(index != -1) {
             mLayoutManager.scrollToPositionWithOffset( index, top);
         }
     }
 
     private void updateFeed() {
 
         RetrofitInterface retrofitInterface = ((MainActivity) getActivity()).getRetrofitInterface();
-        retrofitInterface.getNewsFeed("sessionid=" + getArguments().getString(Constants.SESSION_ID)).enqueue(new Callback<NewsFeedResponse>() {
+        retrofitInterface.getNewsFeed(((MainActivity)getActivity()).getSessionIDHeader()).enqueue(new Callback<NewsFeedResponse>() {
```
Comment (unstablebrainiac) [inline app/src/main/java/app/insti/fragment/FeedFragment.java]: Consider moving MainActivity$getSessionIDHeader() to Helpers

File: app/src/main/java/app/insti/fragment/FeedFragment.java [hunk 2]
```diff
@@ -122,12 +126,10 @@ public void onFailure(Call<NewsFeedResponse> call, Throwable t) {
         });
     }
 
-    private void displayEvents(final List<Event> events) {
-        /* Skip if we're already destroyed */
-        if (getActivity() == null || getView() == null) return;
-
+    /** Initialize the add event fab if the user has permission */
+    private void initFab() {
         if (((MainActivity) getActivity()).createEventAccess()) {
```
Comment (unstablebrainiac) [inline app/src/main/java/app/insti/fragment/FeedFragment.java]: Consider moving MainActivity$createEventAccess() to Helpers

File: app/src/main/java/app/insti/fragment/FeedFragment.java [hunk 3]
```diff
@@ -145,43 +147,49 @@ public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
...
-        });
+        // Initialize adapter
+        if (feedAdapter == null) {
+            feedAdapter = new FeedAdapter(events, this);
+        } else {
+            feedAdapter.setEvents(events);
+            feedAdapter.notifyDataSetChanged();
+        }
+
+        // Initialize adapter if necessary
```
Comment (unstablebrainiac) [inline app/src/main/java/app/insti/fragment/FeedFragment.java]: Initialize RecyclerView if necessary?

File: app/src/main/java/app/insti/adapter/BodyAdapter.java [hunk 4]
```diff
@@ -39,13 +43,28 @@ public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
...
             }
         });
 
         return postViewHolder;
     }
 
+    /**
+     * Open body fragment for a body
+     */
+    private void openBody(Body body) {
```
[the remark this answers is not in the document]
Comment (pulsejet) [reply app/src/main/java/app/insti/adapter/BodyAdapter.java]: Moved to Utils

File: app/src/main/java/app/insti/fragment/FeedFragment.java (hunk 1 above)
Comment (pulsejet) [reply app/src/main/java/app/insti/fragment/FeedFragment.java]: Done

File: app/src/main/java/app/insti/fragment/FeedFragment.java (hunk 2 above)
Comment (pulsejet) [reply app/src/main/java/app/insti/fragment/FeedFragment.java]: Will leave this to a later refactor

File: app/src/main/java/app/insti/fragment/FeedFragment.java (hunk 3 above)
Comment (pulsejet) [reply app/src/main/java/app/insti/fragment/FeedFragment.java]: Haha, right

File: app/src/main/java/app/insti/UpdatableList.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    public void setCache(List<T> mCache) {
+        cache = mCache;
+    }
+
+    /** Update existing or add */
+    public void updateCache(T t) {
+        for (int i = 0; i < cache.size(); i++) {
+            T cachedEvent = cache.get(i);
```
Comment (unstablebrainiac) [inline app/src/main/java/app/insti/UpdatableList.java]: Change variable name to make it generic

File: app/src/main/java/app/insti/Utils.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        return resizeImageUrl(url, 200);
+    }
+
+    public static final String resizeImageUrl(String url, Integer dim) {
+        if (url == null) { return url; }
+        return url.replace("api.insti.app/static/", "img.insti.app/static/" + dim.toString() + "/");
+    }
+
+    /** Update the open fragment */
+    public static final void updateFragment(Fragment fragment, FragmentActivity fragmentActivity) {
```
Comment (unstablebrainiac) [inline app/src/main/java/app/insti/Utils.java]: Use `Context` as the second parameter here, and in subsequent functions. It need not always be a `FragmentActivity`

File: app/src/main/java/app/insti/activity/MainActivity.java [hunk 7]
```diff
@@ -316,22 +313,17 @@ private void openUserFragment(String id) {
...
-        updateFragment(bodyFragment);
+        Utils.openBodyFragment(new Body(id), this);
     }
 
     /** Open the event fragment from the provided id */
     private void openEventFragment(String id) {
-        RetrofitInterface retrofitInterface = getRetrofitInterface();
-        retrofitInterface.getEvent(getSessionIDHeader(), id).enqueue(new EmptyCallback<Event>() {
+        RetrofitInterface retrofitInterface = Utils.getRetrofitInterface();
+        final FragmentActivity self = this;
```
Comment (unstablebrainiac) [inline app/src/main/java/app/insti/activity/MainActivity.java]: Again, `Context` here

File: app/src/main/java/app/insti/Utils.java (hunk 6 above)
Comment (pulsejet) [reply app/src/main/java/app/insti/Utils.java]: `getSupportFragmentManager` exists only on `Fragment Activity`, so in fact it *has* to be one.

File: app/src/main/java/app/insti/activity/MainActivity.java (hunk 7 above)
Comment (pulsejet) [reply app/src/main/java/app/insti/activity/MainActivity.java]: Same as above

File: app/src/main/java/app/insti/UpdatableList.java (hunk 5 above)
Comment (pulsejet) [reply app/src/main/java/app/insti/UpdatableList.java]: Perils of refactoring

File: app/src/main/java/app/insti/Utils.java (hunk 6 above)
Comment (unstablebrainiac) [reply app/src/main/java/app/insti/Utils.java]: Oh. Didn't know this. Cool.

**Pull request merged**